it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,219 @@
+    InstanceInfo instanceInfo =
+        InstanceInfo.newBuilder(instanceId)
+            .setInstanceConfigId(InstanceConfigId.of(projectId, region))
+            .setDisplayName(instanceId.getInstance())
+            .setNodeCount(1)
+            .build();
+    try {
+      instanceAdminClient.createInstance(instanceInfo).get();
+      LOG.info("Successfully created instance.");
+    } catch (ExecutionException | InterruptedException e) {
+      throw new CreateInstanceException("Failed to create instance.", e);
+    }
+  }
+
+  @Override
+  public synchronized void createTable(String statement)
+      throws CreateDatabaseException, IllegalStateException, UpdateDatabaseException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+
+    if (databaseClient != null) {
+      try {
+        databaseAdminClient
+            .updateDatabaseDdl(
+                instanceId.getInstance(),
+                databaseId.getDatabase(),
+                ImmutableList.of(statement),
+                null)
+            .get();
+        LOG.info("Successfully created table.");
+      } catch (ExecutionException | InterruptedException e) {
+        throw new UpdateDatabaseException("Failed to create table.", e);
+      }
+    } else {
+      try {
+        databaseAdminClient
+            .createDatabase(
+                instanceId.getInstance(), databaseId.getDatabase(), ImmutableList.of(statement))
+            .get();
+        databaseClient = spanner.getDatabaseClient(databaseId);
+        LOG.info("Successfully created database with table.");
+      } catch (ExecutionException | InterruptedException e) {
+        throw new CreateDatabaseException("Failed to create database with table.", e);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void insertTableRecord(Mutation tableRecord) throws IllegalStateException {
+    insertTableRecords(ImmutableList.of(tableRecord));
+  }
+
+  @Override
+  public synchronized void insertTableRecords(Iterable<Mutation> tableRecords)
+      throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+    databaseClient.write(tableRecords);
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, String... columnNames)
+      throws IllegalStateException {
+    return readTableRecords(tableId, ImmutableList.copyOf(columnNames));
+  }
+
+  @Override
+  public synchronized ResultSet readTableRecords(String tableId, Iterable<String> columnNames)
+      throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException(
+          "Manager has cleaned up all resources and cannot complete operation.");
+    }
+    return databaseClient.singleUse().read(tableId, KeySet.all(), columnNames);
```

Comment:
Possible problem with the `databaseClient` not being initialized here, because AFAIU it's created only in the `createTable` method. If someone tries to read table records before calling `createTable` this will simply throw an NPE.

If reading a table before creating at least 1 is not supported, at least a better check/error message is needed.

But in fact I don't see why we should have this restriction. I think we can easily have a template that creates the target Spanner tables itself. So an integration test would need to read this data without calling `createTable` to verify the results.